akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -418,23 +422,25 @@ private[http] object HttpServerBluePrint {
         override def onUpstreamFailure(ex: Throwable): Unit = fail(target, ex)
       })
 
+      val shutdown: () ⇒ Unit = () ⇒ completeStage()
+      val httpToNet: ResponseRenderingOutput ⇒ Unit = {
+        case HttpData(b) ⇒ push(toNet, b)
+        case SwitchToWebsocket(bytes, handlerFlow) ⇒
+          push(toNet, bytes)
+          val frameHandler = handlerFlow match {
+            case Left(frameHandler) ⇒ frameHandler
+            case Right(messageHandler) ⇒
+              Websocket.stack(serverSide = true, maskingRandomFactory = websocketRandomFactory, log = log).join(messageHandler)
+          }
+          installHandler(frameHandler)
+          websocketHandlerWasInstalled = true
+      }
+      val wsToNet: ByteString ⇒ Unit = push(toNet, _)
+
       setHandler(toNet, new OutHandler {
-        import akka.http.impl.engine.rendering.ResponseRenderingOutput._
         def onPull(): Unit =
-          if (isHttp) read(fromHttp) {
-            case HttpData(b) ⇒ push(toNet, b)
-            case SwitchToWebsocket(bytes, handlerFlow) ⇒
-              push(toNet, bytes)
-              val frameHandler = handlerFlow match {
-                case Left(frameHandler) ⇒ frameHandler
-                case Right(messageHandler) ⇒
-                  Websocket.stack(serverSide = true, maskingRandomFactory = websocketRandomFactory, log = log).join(messageHandler)
-              }
-              installHandler(frameHandler)
-              websocketHandlerWasInstalled = true
-          }
-          else
-            read(fromWs)(push(toNet, _))
+          if (isHttp) read(fromHttp)(httpToNet, shutdown)
```

Comment:
Will it ever make sense to make this less hard-coded? (Upgrade to more than WS I mean)